File: loader/src/main/java/com/blackmorse/hattrick/promotions/PromotionsCalculator.java
package com.blackmorse.hattrick.promotions;

import com.blackmorse.hattrick.api.worlddetails.model.League;
import com.blackmorse.hattrick.promotions.model.*;
import lombok.Getter;

import java.util.*;
import java.util.stream.Collectors;

@Getter
public class PromotionsCalculator {
    private static final Map<Integer, DivisionDownStrategy> downStrategies = new HashMap<>();

    static {
        downStrategies.put(1, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.FORWARD));
        downStrategies.put(2, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.FORWARD));
        downStrategies.put(3, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.FORWARD));
        downStrategies.put(4, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.FORWARD));
        downStrategies.put(5, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.FORWARD));
        downStrategies.put(6, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.REVERSE));
        downStrategies.put(7, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.REVERSE));
        downStrategies.put(8, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.REVERSE));
        downStrategies.put(9, new DivisionDownStrategy(DownStrategy.REVERSE, DownStrategy.REVERSE));
        downStrategies.put(10, new DivisionDownStrategy(DownStrategy.NONE, DownStrategy.NONE));
    }

    private final Map<Integer, DivisionTeams> divisionTeams = new HashMap<>();
    private final PromotionListsMerger<PromoteTeam, Promotion> merger;// = new PromotionListsMerger<>(Promotion::new);

    public PromotionsCalculator(League league, Integer season) {
        for (int divisionLevel = 1; divisionLevel <= league.getNumberOfLevels() - 1; divisionLevel++) {
            divisionTeams.put(divisionLevel, new DivisionTeams(divisionLevel, downStrategies.get(divisionLevel)));
        }

        divisionTeams.put(league.getNumberOfLevels(), new DivisionTeams(league.getNumberOfLevels(),
                new DivisionDownStrategy(DownStrategy.NONE, DownStrategy.NONE)));

        this.merger = new PromotionListsMerger<>((l1, l2) -> new Promotion(league.getLeagueId(), season, l1, l2));
    }

    public void addPromoteTeam(PromoteTeam promoteTeam) {
        divisionTeams.get(promoteTeam.getTeam().getLeagueUnit().getLevel()).teams.add(promoteTeam);
    }

    private static final Comparator<PromoteTeam> straightComparator =
            Comparator.comparing(PromoteTeam::getPosition)
                .thenComparing(promoteTeam -> -promoteTeam.getPoints())
                .thenComparing(promoteTeam -> -promoteTeam.getDiff())
                .thenComparing(promoteTeam -> -promoteTeam.getScored());

    private static final Comparator<PromoteTeam> reverseComparator = straightComparator.reversed();

    public List<Promotion> calculatePromotionsForDivision(Integer divisionLevel) {
        DivisionDownStrategy downStrategy = divisionTeams.get(divisionLevel).downStrategy;

        List<PromoteTeam> qualifyGoingDownTeams = divisionTeams.get(divisionLevel).teams.stream()
                .filter(promoteTeam -> promoteTeam.getPosition() == 5 || promoteTeam.getPosition() == 6)
                .sorted(straightComparator)
                .peek(promoteTeam -> promoteTeam.promoteType = PromoteType.QUALIFY)
                .collect(Collectors.toList());

        List<PromoteTeam> autoGoingDownTeams = divisionTeams.get(divisionLevel).teams.stream()
                .filter(promoteTeam -> promoteTeam.getPosition() == 7 || promoteTeam.getPosition() == 8)
                .sorted(straightComparator)
                .peek(promoteTeam -> promoteTeam.promoteType = PromoteType.AUTO)
                .collect(Collectors.toList());

        List<PromoteTeam> goingUpTeams = divisionTeams.get(divisionLevel + 1).teams.stream()
                .sorted(straightComparator)
                .limit(qualifyGoingDownTeams.size() + autoGoingDownTeams.size())
                .collect(Collectors.toList());


        List<PromoteTeam> autoGoingUpTeams = goingUpTeams.stream()
                .limit(autoGoingDownTeams.size())
                .sorted(downStrategy.getAutoPromoteStrategy().equals(DownStrategy.FORWARD) ? straightComparator : reverseComparator)
                .peek(promoteTeam -> promoteTeam.promoteType = PromoteType.AUTO)
                .collect(Collectors.toList());

        List<PromoteTeam> qualifyGoingUpTeams = goingUpTeams.stream()
                .skip(autoGoingDownTeams.size())
                .limit(qualifyGoingDownTeams.size())
                .sorted(downStrategy.getQualifyPromoteStrategy().equals(DownStrategy.FORWARD) ? straightComparator : reverseComparator)
                .peek(promoteTeam -> promoteTeam.promoteType = PromoteType.QUALIFY)
                .collect(Collectors.toList());

        List<Promotion> result = new ArrayList<>();

        List<Promotion> autoPromotions = merger.mergeTeams(autoGoingDownTeams, autoGoingUpTeams);
        List<Promotion> qualifyPromotions = merger.mergeTeams(qualifyGoingDownTeams, qualifyGoingUpTeams);

        result.addAll(autoPromotions);
        result.addAll(qualifyPromotions);

        return result;
    }


}

File: akka-loader/src/main/scala/promotions/PromotionsCalculator.scala
package promotions

import chpp.worlddetails.models.League
import loadergraph.promotions.{Auto, DivisionDownStrategy, DivisionTeams, None, PromotionListsMerger, Qualify, Reverse, Straightforward}
import models.clickhouse.PromotionModelCH
import models.stream.StreamTeam
import promotions.PromotionsCalculator.{downStrategies, reverseOrdering, straightOrdering}

object PromotionsCalculator {
  val downStrategies = Map(
    1 -> DivisionDownStrategy(Reverse, Straightforward),
    2 -> DivisionDownStrategy(Reverse, Straightforward),
    3 -> DivisionDownStrategy(Reverse, Straightforward),
    4 -> DivisionDownStrategy(Reverse, Straightforward),
    5 -> DivisionDownStrategy(Reverse, Straightforward),
    6 -> DivisionDownStrategy(Reverse, Reverse),
    7 -> DivisionDownStrategy(Reverse, Reverse),
    8 -> DivisionDownStrategy(Reverse, Reverse),
    9 -> DivisionDownStrategy(Reverse, Reverse),
    10 -> DivisionDownStrategy(None, None),
  )

  private val straightOrdering: Ordering[StreamTeam] = Ordering
    .by[StreamTeam, Int](_.position)
    .orElseBy(- _.points)
    .orElseBy(- _.diff)
    .orElseBy(- _.scored)

  private val reverseOrdering = straightOrdering.reverse

  def calculatePromotions(promotionsCalculator: PromotionsCalculator, levels: Int): List[PromotionModelCH] = {
    if(promotionsCalculator.divisionTeams.isEmpty) return List()
    (1 until levels).flatMap(level => promotionsCalculator.calculatePromotionsForDivision(level)).toList
  }
}

class PromotionsCalculator(league: League, teams: List[StreamTeam]) {
  private val divisionTeams: Map[Int, DivisionTeams] = teams.groupBy(_.leagueUnit.level)
    .map{case (level, teams) => (level, DivisionTeams(level, if (level != league.numberOfLevels ) downStrategies(level) else DivisionDownStrategy(None, None), teams))}


  def calculatePromotionsForDivision(divisionLevel: Int): List[PromotionModelCH] = {
    val downStrategy = divisionTeams(divisionLevel).divisionDownStrategy

    val qualifyGoingDownTeams = divisionTeams(divisionLevel).teams
      .filter(team => team.position == 5 || team.position == 6)
      .sorted(PromotionsCalculator.straightOrdering)

    val autoGoingDownTeams = divisionTeams(divisionLevel).teams
      .filter(team => team.position == 7 || team.position == 8)
      .sorted(PromotionsCalculator.straightOrdering)

    val goingUpTeams = divisionTeams(divisionLevel + 1).teams
      .sorted(PromotionsCalculator.straightOrdering)
      .take(qualifyGoingDownTeams.size + autoGoingDownTeams.size)

    val autoGoingUpTeams = goingUpTeams
      .take(autoGoingDownTeams.size)
      .sorted(if (downStrategy.autoPromoteStrategy == Straightforward) straightOrdering else reverseOrdering)

    val qualifyGoingUpTeams = goingUpTeams
      .slice(autoGoingDownTeams.size, autoGoingDownTeams.size + qualifyGoingDownTeams.size)
      .sorted(if (downStrategy.qualifyPromoteStrategy == Straightforward) straightOrdering else reverseOrdering)

    val autoPromotions = PromotionListsMerger.merge(leagueId = league.leagueId,
      season = league.season - league.seasonOffset,
      upDivisionLevel = divisionLevel,
      promoteType = Auto,
      leftList = autoGoingDownTeams,
      rightList = autoGoingUpTeams
    )

    val qualifyPromotions = PromotionListsMerger.merge(leagueId = league.leagueId,
      season = league.season - league.seasonOffset,
      upDivisionLevel = divisionLevel,
      promoteType = Qualify,
      leftList = qualifyGoingDownTeams,
      rightList = qualifyGoingUpTeams
    )

    autoPromotions ++ qualifyPromotions
  }
}
